### PR TITLE
Attribute Improvements

### PR DESCRIPTION
**Pull request by corindwyer**
### Description of the Change
* Add apis for updating/deleting attributes for jobs, tasks, instance groups, and agent instances.
* Serialize all agent management model updates.
* Move instance type into the instance group data model instead of using a custom attribute.
* Extract agent attributes into component specific attributes

### CODE REVIEW

File: titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobAttributesTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    @Test(timeout = TEST_TIMEOUT_MS)
+    public void testDeleteJobAttributes() throws Exception {
+        JobDescriptor<ServiceJobExt> job = oneTaskServiceJobDescriptor();
+
+        jobsScenarioBuilder.schedule(job, jobScenarioBuilder -> jobScenarioBuilder
+                .template(ScenarioTemplates.jobAccepted())
+                .updateJobAttributes(Collections.singletonMap("a", "1"))
+                .deleteJobAttributes(Collections.singletonList("a"))
```
Comment (tbak) [inline titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobAttributesTest.java]: Asserts missing

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/AgentManagementClient.java [hunk 2]
```diff
@@ -45,7 +47,11 @@
 
     Completable updateInstanceGroupAttributes(InstanceGroupAttributesUpdate attributesUpdate);
 
+    Completable deleteInstanceGroupAttributes(DeleteInstanceGroupAttributesRequest deleteRequest);
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/agent/AgentManagementClient.java]: Use SpringReactor for the new methods

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java [hunk 3]
```diff
@@ -126,6 +130,35 @@ public Response setJobDisruptionBudget(@PathParam("jobId") String jobId,
...
+            sanitizedRequest = request;
+        }
+        return Responses.fromVoidMono(jobManagementClient.updateJobAttributes(sanitizedRequest));
+    }
+
+    @DELETE
+    @ApiOperation("Delete attributes of a job with the specified key names")
+    @Path("/jobs/{jobId}/attributes")
+    public Response deleteJobAttributes(@PathParam("jobId") String jobId,
+                                        @QueryParam("keys") final List<String> keys) {
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java]: Shouldn't this be `String` value with comma separated key list?

Comment (corindwyer) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java]: I have seen lists represented both as delimited strings as well as multiple query params with the same name. I just chose the JaxRS way (multiple query params), but I don't have a strong preference.

File: titus-ext/aws/src/main/java/com/netflix/titus/ext/aws/AwsInstanceCloudConnector.java [hunk 4]
```diff
@@ -442,15 +444,19 @@ private InstanceGroup toInstanceGroup(AutoScalingGroup awsScalingGroup) {
                 isTerminateSuspended = true;
             }
         }
+        Map<String, String> attributes = CollectionsExt.isNullOrEmpty(awsScalingGroup.getTags())
+                ? Collections.emptyMap()
+                : awsScalingGroup.getTags().stream().collect(Collectors.toMap(TagDescription::getKey, TagDescription::getValue));
         return new InstanceGroup(
                 awsScalingGroup.getAutoScalingGroupName(),
                 awsScalingGroup.getLaunchConfigurationName(),
+                DEFAULT_INSTANCE_TYPE,
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-ext/aws/src/main/java/com/netflix/titus/ext/aws/AwsInstanceCloudConnector.java]: This is not available from the ASG call. There is an additional call to get this information from the launch configuration.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java (hunk 3 above)
Comment (tbak) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java]: In Titus REST API we encode multiple keys (job ids, task ids) as delimited strings.

File: titus-api/src/main/java/com/netflix/titus/api/agent/service/AgentManagementService.java [hunk 5]
```diff
@@ -60,19 +57,33 @@
...
      */
     Completable updateInstanceGroupAttributes(String instanceGroupId, Map<String, String> attributes);
 
     /**
-     * Changes attributes of an agent instance.
+     * Deletes attributes of an instance group based on the key names.
+     *
+     * @return AgentManagementException if the instance group is not found
+     */
+    Completable deleteInstanceGroupAttributes(String instanceGroupId, List<String> keys);
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-api/src/main/java/com/netflix/titus/api/agent/service/AgentManagementService.java]: Fixed

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/V3JobOperations.java [hunk 6]
```diff
@@ -72,6 +74,10 @@
 
     Mono<Void> updateJobDisruptionBudget(String jobId, DisruptionBudget disruptionBudget);
 
+    Mono<Void> updateJobAttributes(String jobId, Map<String, String> attributes);
+
+    Mono<Void> deleteJobAttributes(String jobId, List<String> keys);
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/service/V3JobOperations.java]: Fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/AgentCache.java [hunk 7]
```diff
@@ -44,6 +46,12 @@
 
     Completable updateAgentInstanceStore(AgentInstance agentInstance);
 
+    Single<AgentInstanceGroup> getAndUpdateInstanceGroupStore(String instanceGroupId, Function<AgentInstanceGroup, AgentInstanceGroup> function);
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/AgentCache.java]: Fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/DefaultAgentCache.java [hunk 8]
```diff
@@ -44,8 +45,10 @@
 import rx.Completable;
 import rx.Observable;
 import rx.Scheduler;
+import rx.Single;
 import rx.Subscription;
 import rx.functions.Action0;
+import rx.functions.Func0;
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/DefaultAgentCache.java]: Fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/InstanceCache.java [hunk 9]
```diff
@@ -349,22 +347,18 @@ private void updateCache(InstanceGroup updatedInstanceGroup, List<Instance> upda
...
-                ATTR_INSTANCE_TYPE, launchConfiguration.getInstanceType()
-        );
+    private InstanceGroup addInstanceType(InstanceGroup instanceGroup, InstanceLaunchConfiguration launchConfiguration) {
         return instanceGroup.toBuilder()
-                .withAttributes(newAttributes)
+                .withInstanceType(launchConfiguration.getInstanceType())
                 .build();
     }
 
     private InstanceGroup decorate(InstanceGroup newInstanceGroup, InstanceGroup oldInstanceGroup) {
```
[the remark this answers is not in the document]
Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/agent/service/cache/InstanceCache.java]: Fixed

File: titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobAttributesTest.java (hunk 1 above)
Comment (corindwyer) [reply titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobAttributesTest.java]: the assert logic exists in update and delete job methods

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java (hunk 3 above)
Comment (corindwyer) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/v3/rest/JobManagementResource.java]: I will change it to be consistent.

File: titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/action/task/BasicJobActions.java [hunk 10]
```diff
@@ -75,6 +80,42 @@ public static ChangeAction updateJobDisruptionBudget(ReconciliationEngine<JobMan
                 });
     }
 
+    /**
+     * Update the attributes of a job. This will either create new attributes or replacing an existing ones with the same key.
+     */
+    public static ChangeAction updateJobAttributes(ReconciliationEngine<JobManagerReconcilerEvent> engine, Map<String, String> attributes, JobStore jobStore, CallMetadata callMetadata) {
```
Comment (tbak) [inline titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/action/task/BasicJobActions.java]: `CallMetadata` not set

Comment (corindwyer) [reply titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/common/action/task/BasicJobActions.java]: Fixed

**Pull request merged**